### PR TITLE
Passing process id of parent process for monitoring.

### PR DESCRIPTION
**Pull request by harshjain2**
Passing process id of parent process for monitoring.

### CODE REVIEW

File: src/Microsoft.TestPlatform.CrossPlatEngine/Helpers/Interfaces/IProcessHelper.cs [hunk 1]
```diff
@@ -56,5 +57,17 @@ internal interface IProcessHelper
...
+        /// The wait for parent process exit async.
+        /// </summary>
+        /// <param name="parentProcessId">
+        /// The parent process id.
+        /// </param>
+        /// <param name="requestingEntity"></param>
+        /// <returns>
+        /// The <see cref="Task"/>.
+        /// </returns>
+        Task WaitForParentProcessExitAsync(int parentProcessId, string requestingEntity);
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Helpers/Interfaces/IProcessHelper.cs]: There isn't a concept of `Parent` process in this API.

I am not sure if it's the right approach to dedicate an entire thread to wait for process exit. Instead we should just set a ExitedCallback and let the consumers take an action on that.

```
IProcessHelper.SetExitCallback(int processId, Action<Process> callback)

// In testhost or datacollector, use wait handles as required
```

**Review by codito — CHANGES_REQUESTED** [review]
Request changes to API

File: src/Microsoft.TestPlatform.CrossPlatEngine/Helpers/ProcessHelper.cs [hunk 2]
```diff
@@ -117,19 +117,16 @@ public bool TryGetExitCode(Process process, out int exitCode)
         }
 
         /// <inheritdoc/>
-        public Task WaitForParentProcessExitAsync(int parentProcessId, string requestingEntity)
+        public void SetExitCallback(int parentProcessId, Action callbackAction)
         {
             var parentProcessExitedHandle = new AutoResetEvent(false);
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Helpers/ProcessHelper.cs]: Not used

**Review by codito — APPROVED** [review]

File: src/Microsoft.TestPlatform.CrossPlatEngine/Helpers/ProcessHelper.cs [hunk 3]
```diff
@@ -113,5 +115,18 @@ public bool TryGetExitCode(Process process, out int exitCode)
             exitCode = 0;
             return false;
         }
+
+        /// <inheritdoc/>
+        public void SetExitCallback(int parentProcessId, Action callbackAction)
+        {
+            var parentProcessExitedHandle = new AutoResetEvent(false);
```
Comment (smadala) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Helpers/ProcessHelper.cs]: `parentProcessExitedHandle` not used anymore.

File: src/datacollector/Program.cs [hunk 4]
```diff
@@ -119,5 +135,23 @@ private static void WaitForDebuggerIfEnabled()
...
+                // Wait for the connection to the sender and start processing requests from sender
+                if (requestHandler.WaitForRequestSenderConnection(ClientListenTimeOut))
+                {
+                    requestHandler.ProcessRequests();
+                }
+                else
+                {
+                    EqtTrace.Info("DataCollector: RequestHandler timed out while connecting to the Sender.");
+                    requestHandler.Close();
+                    throw new TimeoutException();
```
Comment (smadala) [inline src/datacollector/Program.cs]: Add message to exception, better if we use Custom exception.

File: src/datacollector/Program.cs [hunk 5]
```diff
@@ -62,6 +70,19 @@ private static void Run(string[] args)
             var argsDictionary = CommandLineArgumentsHelper.GetArgumentsDictionary(args);
             var requestHandler = DataCollectionRequestHandler.Create(new SocketCommunicationManager(), new MessageSink());
 
+            // Attach to exit of parent process
+            string optionValue;
+            var parentProcessId = argsDictionary.TryGetValue(ParentProcessArgument, out optionValue) ? int.Parse(optionValue) : 0;
```
Comment (smadala) [inline src/datacollector/Program.cs]: Parsing Int values can be moved to CoreUtilities as it is used by both testhost and datacollector.

**Review by smadala — APPROVED** [review]

**Pull request merged**